### ISSUE TITLE
Facebook stops responding after a while

### ISSUE THREAD

**Issue by beatlevic** [DracoBlue/node-facebook-client#15]
We are also using the Facebook client, and we seem to be experiencing the same issue as reported by Ran (reported on your weblog). We request the facebook api every 30 seconds for an update, and do that for multiple users (separately in their own respective loops). After a while, facebook stops responding, and all requests freeze and return nothing. Only after a restart of node everything works fine again, until facebook stops responding again. Do you have any idea? Could it be caused by the fact that every request is done with the same agent?
I don't believe we are hitting some Facebook limit, otherwise a restart wouldn't help. Any help here is greatly appreciated.

The freezing issue could be related to an https issue in node, see https://github.com/joyent/node/issues/728 which is fixed in node after v0.4.12, but this is just a wild guess.

**Comment by DracoBlue** [comment]
This is the code, which does the request:

https://github.com/DracoBlue/node-facebook-client/blob/master/lib/facebook-client/FacebookClient.js#L18

It actually does not handle the case of an timeout on it's own, but waits for an error event. Can you confirm that those requests are made, but not closed?

What version of nodejs are you using?

**Comment by beatlevic** [comment]
We wrapped it in a timeout, and at some point we only get timeouts:

```
facebookClient.getSessionByAccessToken(self.auth.token)(function(fbSession) {
  var timedOut = false;
  var callTimeout = setTimeout(function() {
    timedOut = true;
    return callback(new Error("We had some problems contacting the Facebook API"));
  }, 18000);
  fbSession.graphCall('/me/home', params)(function(response) {
    if (timedOut) return null;
    clearTimeout(callTimeout);
    // do stuff with response
```

So far it has been hard to reproduce. It seemed we could reproduce it once or twice by switching from one wireless network to another (or cable). And only after a restart of node facebook would stop producing timeouts.

We use node v0.4.12

**Comment by DracoBlue** [comment]
I guess this would leave the "dead" request still in the FacebookClient.

Can you put your `setTimeout`-handling into the `FacebookClient`'s `doRequest` method and `request.abort()` (remember to try+catch) the request in case of `timedOut` is `true`?

**Comment by beatlevic** [comment]
Makes sense to have the timeout handling in the actual facebookClient's doRequest. I'll see what I can do, and try to roll it out afterwards to see if it still generates the issue.

**Comment by beatlevic** [comment]
We found out that by pointing api.facebook.com and graph.facebook.com to nowhere in our hosts file, we would get the timeout events just like you would expect, only when we stopped pointing the urls to nowhere, the timeouts would continue (and would need a restart).
When we moved the timeout logic into the client, it seemed to have fixed this issue. doRequest() now looks like this:

```
function doRequest(host, port, path, secure, method, data, parser) {
return function(cb) {
    var protocol = http;
    if(secure) protocol = https;
    var options = {
        host: host,
        port: port,
        path: path,
        method: method || 'GET'
    };
    var timedOut = false;
    var callTimeout = null;

    var request = protocol.request(options, function(response){
        response.setEncoding("utf8");

        var body = [];

        response.on("data", function (chunk) {
            body.push(chunk);
        });

        response.on("end", function () {
            if (timedOut) return null;
            clearTimeout(callTimeout);
            cb(null, parser(body.join("")));
        });
    });

    callTimeout = setTimeout(function() {
          console.debug("[FacebookClient] TIMEOUT");
          timedOut = true;
          request.abort();
          return cb(new Error("Facebook API timeout"));
        }, 10000);


    if (data != null) {
        request.write(querystring.stringify(data))
    }
    request.end();
};
};
```

The callback now contains err and response. I set the timeout on 10 sec, but it would perhaps be nice to have that configurable. If you can confirm it, and patch the code, then we will role it out in production to see if it fixes our problem.

**Comment by DracoBlue** [comment]
Nice, I'll add it to the next release today.
